### DISCUSSION TITLE
How about gitops cli tool ?

### DISCUSSION THREAD

**Discussion by vijaySamanuri** [open-gitops/project#78] [General]
We have several environment stages at our place but `gitops` doesn't cover the promotion part among these stages.
As a workaround we adopted "branch per environment" strategy and rotating the PR to the other environment but these all are manual steps and when we script this to CI system its all non standard i.e different CI system requires different way of doing it for example github actions requires different way of doing it, gitlab plugin different way.

I think it is beneficial to have a standard cli tool something like `gitops` which deals with all the manual steps during the gitops deployment process. This tool just interacts with the underlying git tool and wraps the gitops specific git commands.

for example:

`gitops init hld --for-app=guestbook` would initiate the HLD repository if not present
`gitops env add stage` would create a `stage` branch in HLD repository
`gitops raisePR prod --from=stage` raises a PR from `stage` branch to `prod` branch
`gitops rollback prod`
etc

**Comment by vijaySamanuri** [comment]
@scottrigby Scott Rigby any thoughts on this? Tell me in case if it is irrelevant and out of scope ?

**Reply by lloydchang** [reply]
@vijaySamanuri @ams0 @tonit @scottrigby

I believe @vijaySamanuri's request is out of scope because:

`gitops raisePR` would mostly be in conflict with a goal from the [GitOps Working Group Charter](https://docs.google.com/document/d/11EZfvB2FFI837nMmArnyv-wizsIJvc-4_xdgfoUXF4o/):
> **Goals**
> • Project/vendor neutral.

-----

My reasons are:

1. The only Git pull request tooling that is vendor neutral is:
```
git git-request-pull
```
described at https://git-scm.com/docs/git-request-pull

In practice, `git git-request-pull` is project specific, such as https://www.kernel.org/doc/html/latest/maintainer/pull-requests.html

2. Most pull request tooling have vendor bias, often driven by market share and rate of adoption.

Even alternatives to pull requests, such as Gerrit code reviews, references GitHub because of popularity.

For example:
> **[Basic Gerrit Walkthrough — For GitHub Users](https://gerrit-review.googlesource.com/Documentation/intro-gerrit-walkthrough-github.html)**
> **Note:** This document aims to provide a concise description of the core principles of code review in Gerrit for people that were previously using Pull Requests on Github or similar concepts. Nothing in this document is meant to state that one or the other might be better, but only aims to help new users understand Gerrit more readily. We use Github as the point of comparison since it seems to be the most popular service.

3. While Git facilitates GitOps, Git is optional.

If desired, other revision control systems, such as Mercurial and Subversion, can be used instead of Git.

-----

Thank you @vijaySamanuri @ams0 @tonit @scottrigby

**Comment by ams0** [comment]
Isn't the whole point of `GitOps` to manage infrastructure and applications via `git`? Is it not then that you really only need:

```console
git init
git branch staging
gh pr create #Github CLI
git reset --soft HEAD~1
```
?

**Reply by lloydchang** [reply]
The objective of GitOps is convergence to a declarative state.

While Git facilitates GitOps, Git is optional.

This relates to https://github.com/gitops-working-group/gitops-working-group/discussions/76#discussioncomment-633924

Thank you @ams0

**Comment by scottrigby** [comment]
I'd suggest re-opening this as a discussion topic now that #36 is set up. I'd convert it for you, but for the moment we can't due to https://github.com/github/feedback/discussions/2924#discussioncomment-254592.

**Reply by scottrigby** [reply]
Converted this issue to a Discussion proposal.

**Comment by tonit** [comment]
I agree with @ams0 that "pull request driven operations" really seem to be the theme around gitops. 

Still i can imagine that imperative CLI tools have its place when carefully crafting/shaping pull requests. 

Similar to tooling around GitFlow you want to automate most (boring) tasks such as templating a (common) change, testing,  managing potential dependencies between PRs etc.

WDYT?

**Reply by lloydchang** [reply]
@tonit If your focus is on pull requests, then this relates to https://github.com/gitops-working-group/gitops-working-group/discussions/76#discussioncomment-633924

The objective of GitOps is convergence to a declarative state.

If your focus is on operations, then are the following CLIs sufficient?

Argo CD CLI:
• https://argoproj.github.io/argo-cd/getting_started/#creating-apps-via-cli

Flux v1 CLI:
• https://docs.fluxcd.io/en/1.22.1/references/fluxctl/

Flux v2 CLI:
• https://toolkit.fluxcd.io/cmd/flux/

Keptn CLI:
• https://keptn.sh/docs/0.8.x/operate/install/

Thank you @tonit

**Comment by scottrigby** [comment]
There is some new thinking on this, and it was discussed in the most recent GitOps WG meeting. There is some talk about this as a shared CLI command, where support for specific tools would be added through plugins.

The idea is this would live in the OpenGitOps project. See open-gitops/project#77 

This is not an answer yet, just a note that there is some movement here, and will probably move to https://github.com/open-gitops/project/discussions. Keeping this open until it's started there.

**Reply by lloydchang** [reply]
Thank you @scottrigby

When your time permits, here is a wall of text 😄

I tried looking up July's recording at https://www.youtube.com/channel/UCI6iqYuuI4gZuOCZaks5i1g/videos to learn what was discussed, but I couldn't find any uploaded recordings after May.

I did find meeting notes at https://docs.google.com/document/d/1hxifmCdOV5_FbKloDJRWZQHq0ge-trXJKF-BgV4wHVk/edit and I see that OCI was mentioned as an example on July 21st.  Practically-speaking, OCI's cornerstone was runC donated by Docker, summarized at https://opencontainers.org/about/overview/

Since Flux [didn't integrate](https://github.com/fluxcd/flux/blob/master/README.md#flux
) into [GitOps Engine](https://github.com/argoproj/gitops-engine), Flux and Argo are two different implementations of GitOps.

I don't see any issue with multiple implementations.

As for CLI examples, growth, and evolutions: I suppose one can try using [sed](https://en.wikipedia.org/wiki/Sed) for UNIX, [Edlin](https://en.wikipedia.org/wiki/Edlin) for DOS, but those CLIs didn't serve the broader needs that vi and emacs cover.  vi and emacs have co-existed for decades as different implementations of editors that serve needs that CLIs don't.  They each serve their own purpose based on people's preferences.

Although I feel and think that GitOps ecosystem has grown and evolved beyond _just_ a `gitops` CLI, let's go back to the topic of CLI for the sake of this discussion open-gitops/gitops-working-group#76.

A number of questions come to mind:
1. Curiously, who would maintain a CLI?
2. Has GitOps WG considered [bash shell aliases](https://www.cyberciti.biz/tips/bash-aliases-mac-centos-linux-unix.html) to satisfy the immediate needs discussed above?
3. Does GitOps WG prefer typing long-form `gitops` instead of shell aliases such as `g` for `git`, `f` for `fluxctl`, `a` for `argocd`, `k` for `kubectl`, etc.?
4. If not [bash shell aliases](https://www.cyberciti.biz/tips/bash-aliases-mac-centos-linux-unix.html), then which programming language would GitOps WG write the `gitops` CLI in?
5. Which target operating systems would the `gitops` CLI compilation target?
6. Is this turning into [over-engineering](https://en.wikipedia.org/wiki/Overengineering)?
7. Is this conversation trending into [bikeshedding](https://thedecisionlab.com/biases/bikeshedding/)?
8. The point I'm getting at is that the best software code can be code that doesn't need to be written in the first place.
9. I believe the problem space can be solved in more important ways, such as education and training about using software in [GitOps Landscape](https://github.com/open-gitops/documents/pull/34) _even over_ creating and maintaining _just_ a `gitops` CLI.

Thank you for reading 😄

**Comment by lloydchang** [comment]
@vijaySamanuri wrote:
> `gitops init hld --for-app=guestbook` would initiate the HLD repository if not present

---

While I was reviewing:
> [GitOps Working Group - Overview and Invitation.](https://archive.fosdem.org/2021/schedule/event/gitops_working_group/)

[Video](https://archive.fosdem.org/2021/schedule/event/gitops_working_group/) referenced https://github.com/podtato-head/podtato-head/ for the purpose of 📨🚚 CNCF App Delivery SIG Demo.

From https://github.com/podtato-head/podtato-head/blob/main/README.md
> **Project _pod_ tato Head**
> Podtato-head demonstrates cloud-native application delivery scenarios using many different tools and services. It is intended to help application delivery support teams test and decide which mechanism(s) to use.

There is an entry:
> [GitOps deployment via Gimlet CLI](https://github.com/podtato-head/podtato-head/blob/main/delivery/gimlet/README.md)

Perhaps `gimlet gitops bootstrap` described at https://gimlet.io/gimlet-cli/bootstrap-gitops-with-gimlet/ has similar mechanics that you expect from `gitops init`?

From your perspective, would both `gitops init` and `gimlet gitops boostrap` operate at the same level of abstraction?

WDYT?

Thank you!
